### ISSUE TITLE
about retry queues, what they are and why they exist.

### ISSUE THREAD

**Issue by petersilva** [MetPX/sarracenia#620]
in the early days of Sarracenia, there were no retry_queues.  There were simply the queues, and when one was having difficulty processing a file, the sarracenia subscriber just kept working on delivering what it had already downloaded the corresponding messages for. Why were retry queues built?

issues:

Queues for long Outages are too big for a broker
----------------------------------------------------------------

when a remote server we are sending to is down, and they are subscribed to a lot of messages, then the queue will build up quickly.  After a few hours, a feed to a storm prediction centre, for example the backlog of messages and corresponding files will be hundreds of thousands, and rapidly climb to millions,  the climat archive, for example, had multiple 7 day outages. Many clients want the data they missed, so the easiest solution is to let the queue build up, 

It is a general tenet of all message passing systems that queues are extremely undesirable.  The preferred status of a queueing system is that all messages are delivered to interested parties and processed by them so that as few as possible are queued at any one point. Rabbitmq literature warns about large queues, and the performance of a broker is affected by the total number of items in all queues on a given broker.  So when a single user suffers an outage, the queue will build up for that user, and all users will see a performance drop.

At any event, when the system is returned the service, the strategy is that, we just what they didn't get in the order it was queued for them, which leads to the second issue:

After long outages, recent data is more important than old data
------------------------------------------------------------------------------------

2 day old weather data is no long weather, it has passed into climatology.  Many clients are in the real-time weather business where the latest data is the most important.  Just having a standard queue build up and then seeing a "catchup" period that could be hours long when our connection to the client is bandwidth limited, just prolongs the outage for them.   What the client wants is for the current data to be fed, and, when there are gaps in current data (so that we are caught up.) then send some old data, but only as long as we aren't blocking any new data.

Some failures are data specific
-----------------------------------------

We have cases where a particular file cannot be successfully transferred.  Reasons are such as that the file exists on the destination, and we are not allowed to delete it, or the target directory is not writable for the sending process.  So one file, or one in a thosand, needs to be retried.  If we just relied on the standard FIFO queue, we just get stuck at this point trying to send this file over and over again... and we never send anything else... This is what Sundew does.  Sarra tries three times, then puts it in a retry queue.
The rest of the data still gets delivered, and we check once in a while, whether delivering the problem file will work later.

How the retry queues work
------------------------------------

The retry queues, as built, store the failures in a sequential file on disk. Sr3 reads from it's input queue, and then if it finds that there is some room (we received less messages than our batch size) it will read some messages from the retry queue into the incoming list of messages to process (worklist.incoming data structure in sr3.)

Then it tries to download/send/process things, the so called "work" phase of the flow algorithm.  if the operation fails, then corresponding item is put onto a new retry queue (not the same one that we are pulling from.) and processing of other messages continues normally.

Every housekeeping interval (5 or 10 minutes, usually), the retry queues are read:
   * expired messages are removed
   * new messages added to the candidates for retry.
  
so:

*  a retry of a failed transfer will never be attempted within one housekeeping interval. only after one full interval has passed.
* retries are lower priority than real-time data as the worklist.incoming queue is always filled with new messages before checking the retry queue.
* clients never get stuck stubbornly trying to retry a failed operation, and do not retry it, even if they have time, more than once per housekeeping interval.

**Comment by petersilva** [comment]
DiskQueue Implementation Details:
-----------------------------------------------

The above algorithm is visible in the sarracenia/flowcb/retry.py ... it makes use of DiskQueue.py.
Within a housekeeping interval:
*  .retry ... the list of candidate messages to pick from to attempt retries.
* .retry.new ... failures that happenned in this interval to be added to the candidate list next interval.  (written to with append.)
* .hk ... the list being built during the housekeeping interval, eventually to replace.retry.

All I/o is append to end of files, no random i/o is used at any point for performance reasons.

**Comment by petersilva** [comment]
Note, there is also a historic implementation, an earlier attempt at dealing with these problems:  save/restore.  with save, one would explicitly read files from an AMQP queue to empty it, and write the corresponding messages to a .json formatted file.

**Restore** would read files from the json file and treat them like a queue.  Once the file was emptied, it would return to processing the normally configured AMQP queue.  It did solve the initial problem of having rabbitmq queue grow too large during failures. However:

Problems:

* processes data in order (oldest first... bad for most clients.)
* does not work around file specific issues, gets stuck on bad data.
* requires complext manual intervention by the analyst to put the subscriber in save mode during the outage, and again to apply restore mode afterward.

The retry module does a better job, and does it without analyst intervention.

**Comment by petersilva** [comment]
All this to say that putting retry queues on brokers will likely not work.  Need something higher capacity that won't slow down other traffix. I'm thinking something like memcached is a better candidate for a shared retry and/or duplicate suppression queue.

**Comment by G-regL** [comment]
> All this to say that putting retry queues on brokers will likely not work. Need something higher capacity that won't slow down other traffic. I'm thinking something like memcached is a better candidate for a shared retry and/or duplicate suppression queue.

Memcached has many uses, but I don't think this is one of them. Evictions can happen any time the memory limits are reached, and there's no persistence across service restarts.

IMO, you'd be much better served with Redis, using Lists in particular. 

- As fast as Memcached
- Supports replication/HA
- Can push/pop to the head or tail of the list, meaning you could use it as a LIFO queue (addresses the use case above where clients probably don't want to process the oldest data first, but rather the newest)

**Comment by MagikEh** [comment]
You'd need two different queues that get joined on a maintenance interval to avoid the old sundew bug where it'd get stuck retrying on a single file,, failing, adding it back to the queue, pulling from the queue, ad-infinitum..

**Comment by G-regL** [comment]
That's a great point!!

Could just create "retry" counter keys for each file that increments each time it does. If the counter is above some value, it cleans up all the keys for that file and moves on.

**Comment by petersilva** [comment]
OK have the implementation, but things we still need:

* add CI/CD actions that run redisqueue throught the flow tests.
* add documentation of use of the redisqueue.
